### PR TITLE
API Changes

### PR DESCRIPTION
**Pull request by prydie**
Started work on making the API more idiomatic prior to v0.2 release. There are still a number of changes to make but feedback at this point would be great.

Includes:
 - Updating dependencies to Kubernetes 1.10.0
 - Regenerating all generated code incl. rename `Mysql -> MySQL`

Resolves: #119

### CODE REVIEW

File: pkg/apis/mysql/v1alpha1/types.go [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+	SecretRef *corev1.LocalObjectReference `json:"secretRef,omitempty"`
+
+	// ConfigRef allows a user to specify a custom configuration file for MySQL.
+	// +optional
+	ConfigRef *corev1.LocalObjectReference `json:"configRef,omitempty"`
+
+	// SSLSecretRef allows a user to specify custom CA certificate, server certificate
+	// and server key for group replication SSL
+	// +optional
+	SSLSecretRef *corev1.LocalObjectReference `json:"sslSecretRef,omitempty"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: `s/SSLSecretRef/RootPasswordSecret/g`

File: pkg/apis/mysql/v1alpha1/types.go [hunk 2]
```diff
@@ -0,0 +1,419 @@
...
+	BackupVolumeClaimTemplate *corev1.PersistentVolumeClaim `json:"backupVolumeClaimTemplate,omitempty"`
+
+	// If defined, we use this secret for configuring the MYSQL_ROOT_PASSWORD
+	// If it is not set we generate a secret dynamically
+	// +optional
+	SecretRef *corev1.LocalObjectReference `json:"secretRef,omitempty"`
+
+	// ConfigRef allows a user to specify a custom configuration file for MySQL.
+	// +optional
+	ConfigRef *corev1.LocalObjectReference `json:"configRef,omitempty"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: `s/ConfigRef/Config/g`?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 3]
```diff
@@ -0,0 +1,419 @@
...
+
+	// BackupVolumeClaimTemplate allows a user to specify a volume to temporarily store the
+	// data for a backup prior to it being shipped to object storage.
+	// +optional
+	BackupVolumeClaimTemplate *corev1.PersistentVolumeClaim `json:"backupVolumeClaimTemplate,omitempty"`
+
+	// If defined, we use this secret for configuring the MYSQL_ROOT_PASSWORD
+	// If it is not set we generate a secret dynamically
+	// +optional
+	SecretRef *corev1.LocalObjectReference `json:"secretRef,omitempty"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: `s/SecretRef/RootPasswordSecret/`?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 4]
```diff
@@ -0,0 +1,419 @@
...
+
+	// BaseServerID defines the base number used to create uniq server_id for MySQL instances in a cluster.
+	// The baseServerId value need to be in range from 1 to 4294967286
+	// If ommited in the manifest file, or set to 0, defaultBaseServerID value will be used.
+	BaseServerID uint32 `json:"baseServerId,omitempty"`
+
+	// MultiMaster defines the mode of the MySQL cluster. If set to true,
+	// all instances will be R/W. If false (the default), only a single instance
+	// will be R/W and the rest will be R/O.
+	MultiMaster bool `json:"multiMaster,omitempty"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Add comment and link to why multi-master is probably not what users want.

File: pkg/apis/mysql/v1alpha1/types.go [hunk 5]
```diff
@@ -0,0 +1,419 @@
...
+	ClusterPhaseFailed,
+	ClusterPhaseUnknown,
+}
+
+// ClusterStatus defines the current status of a MySQL cluster
+// propagating useful information back to the cluster admin
+type ClusterStatus struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata"`
+	Phase             ClusterPhase `json:"phase"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Replace with conditions?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 6]
```diff
@@ -0,0 +1,419 @@
...
+	Errors            []string     `json:"errors"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Remove?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 7]
```diff
@@ -0,0 +1,419 @@
...
+	Name string `json:"name"`
+	// Databases are the databases to backup.
+	Databases []string `json:"databases"`
+}
+
+// BackupStorageProvider defines the configuration for storing a MySQL backup to a storage service.
+// The generation of the backup is configured in the Executor configuration.
+type BackupStorageProvider struct {
+	// Name denotes the type of storage provider that will store and retrieve the backups,
+	// e.g. s3, oci-s3-compat, aws-s3, gce-s3, etc.
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: s3 only currently

File: pkg/apis/mysql/v1alpha1/types.go [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+	Name string `json:"name"`
+	// SecretRef is a reference to the Kubernetes secret containing the configuration for uploading
+	// the backup to authenticated storage.
+	SecretRef *corev1.LocalObjectReference `json:"secretRef,omitempty"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: `s/SecretRef/Secret/`?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 9]
```diff
@@ -0,0 +1,419 @@
...
+// BackupOutcome describes the location of a MySQL Backup
+type BackupOutcome struct {
+	// Location is the Object Storage network location of the Backup.
+	Location string `json:"location"`
+}
+
+// BackupStatus captures the current status of a MySQL backup.
+type BackupStatus struct {
+	// Phase is the current life-cycle phase of the Backup.
+	Phase BackupPhase `json:"phase"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Replace with conditions?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 10]
```diff
@@ -0,0 +1,419 @@
...
+
+	// Outcome holds the results of a successful backup.
+	Outcome BackupOutcome `json:"outcome"`
+
+	// TimeStarted is the time at which the backup was started.
+	TimeStarted metav1.Time `json:"timeStarted"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: These can be folded into conditions I think?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 11]
```diff
@@ -0,0 +1,419 @@
...
+
+	// BackupSchedulePhaseFailedValidation means the backup schedule has failed
+	// the controller's validations and therefore will not trigger backups.
+	BackupSchedulePhaseFailedValidation BackupSchedulePhase = "FailedValidation"
+)
+
+// ScheduleStatus captures the current state of a MySQL backup schedule.
+type ScheduleStatus struct {
+	// Phase is the current phase of the MySQL backup schedule.
+	Phase BackupSchedulePhase `json:"phase"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Replace with conditions?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 12]
```diff
@@ -0,0 +1,419 @@
...
+	metav1.ListMeta `json:"metadata"`
+
+	Items []BackupSchedule `json:"items"`
+}
+
+// RestoreSpec defines the specification for a restore of a MySQL backup.
+type RestoreSpec struct {
+	// ClusterRef is a refeference to the Cluster to which the Restore
+	// belongs.
+	ClusterRef *corev1.LocalObjectReference `json:"clusterRef"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: `s/ClusterRef/Cluster/`?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 13]
```diff
@@ -0,0 +1,419 @@
...
+
+	// BackupRef is a reference to the Backup object to be restored.
+	BackupRef *corev1.LocalObjectReference `json:"backupRef"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: `s/BackupRef/Backup/`?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 14]
```diff
@@ -0,0 +1,419 @@
...
+	RestorePhaseComplete RestorePhase = "Complete"
+
+	// RestorePhaseFailed means the Restore has terminated with an error.
+	RestorePhaseFailed RestorePhase = "Failed"
+)
+
+// RestoreStatus captures the current status of a MySQL restore.
+type RestoreStatus struct {
+	// Phase is the current life-cycle phase of the Restore.
+	Phase RestorePhase `json:"phase"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Replace with conditions?

File: pkg/apis/mysql/v1alpha1/types.go (hunk 1 above)
Comment (simonlord) [reply pkg/apis/mysql/v1alpha1/types.go]: s/SSLSecretRef/SSLSecret/g but i'm guessing that's what you meant ;)

File: pkg/apis/mysql/v1alpha1/types.go (hunk 8 above)
Comment (simonlord) [reply pkg/apis/mysql/v1alpha1/types.go]: Should this be `AuthSecret` or similar? `Secret` on it's own is a bit too generic in k8s world i'd say?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 15]
```diff
@@ -0,0 +1,400 @@
...
+
+	// NodeSelector is a selector which must be true for the pod to fit on a node.
+	// Selector which must match a node's labels for the pod to be scheduled on that node.
+	// More info: https://kubernetes.io/docs/concepts/configuration/assign-pod-node/
+	// +optional
+	NodeSelector map[string]string `json:"nodeSelector,omitempty"`
+
+	// If specified, affinity will define the pod's scheduling constraints
+	// +optional
+	Affinity *corev1.Affinity `json:"affinity,omitempty"`
```
Comment (jhorwit2) [inline pkg/apis/mysql/v1alpha1/types.go]: Affinity breaks PVC/STS. You sure you want this prior to topology aware CSI?

File: pkg/apis/mysql/v1alpha1/types.go [hunk 16]
```diff
@@ -0,0 +1,400 @@
...
+
+	// StorageProvider is the configuration of where and how backups should be stored.
+	StorageProvider *BackupStorageProvider `json:"storageProvider"`
+
+	// Cluster is a reference to the Cluster to which the Backup belongs.
+	Cluster *corev1.LocalObjectReference `json:"cluster"`
+
+	// AgentScheduled is the agent hostname to run the backup on.
+	// TODO(apryde): ScheduledAgent (*corev1.LocalObjectReference)?
+	AgentScheduled string `json:"agentscheduled"`
```
Comment (jhorwit2) [inline pkg/apis/mysql/v1alpha1/types.go]: This should be 

> ScheduledAgent string `json:"scheduledAgent"`

File: pkg/apis/mysql/v1alpha1/types.go [hunk 17]
```diff
@@ -0,0 +1,400 @@
...
+}
+
+// BackupExecutor represents the configuration of the tool performing the backup. This includes the tool
+// to use, and, what database and tables should be backed up.
+// The storage of the backup is configured in the relevant Storage configuration.
+type BackupExecutor struct {
+	// Name of the tool performing the backup, e.g. mysqldump.
+	Name string `json:"name"`
+	// Databases are the databases to backup.
+	Databases []string `json:"databases"`
```
Comment (jhorwit2) [inline pkg/apis/mysql/v1alpha1/types.go]: Do you ever see this needing more fields under database? Like the table? If so i'd recommend using a struct with a name field.

File: pkg/apis/mysql/v1alpha1/types.go [hunk 18]
```diff
@@ -0,0 +1,400 @@
...
+}
+
+// BackupStorageProvider defines the configuration for storing a MySQL backup to a storage service.
+// The generation of the backup is configured in the Executor configuration.
+type BackupStorageProvider struct {
+	// Name denotes the type of storage provider that will store and retrieve the backups.
+	// Currently only supports "S3" denoting a S3 compatiable storage provider.
+	Name string `json:"name"`
```
Comment (jhorwit2) [inline pkg/apis/mysql/v1alpha1/types.go]: Seems like it should be a const.

File: pkg/apis/mysql/v1alpha1/types.go (hunk 9 above)
Comment (jhorwit2) [reply pkg/apis/mysql/v1alpha1/types.go]: +1

https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#typical-status-properties

> Newer API types should use conditions instead. 

You should avoid using phase as it's a deprecated pattern in core.

File: pkg/apis/mysql/v1alpha1/types.go [hunk 19]
```diff
@@ -0,0 +1,400 @@
...
+type RestoreSpec struct {
+	// Cluster is a refeference to the Cluster to which the Restore
+	// belongs.
+	Cluster *corev1.LocalObjectReference `json:"cluster"`
+
+	// Backup is a reference to the Backup object to be restored.
+	Backup *corev1.LocalObjectReference `json:"backup"`
+
+	// AgentScheduled is the agent hostname to run the backup on
+	AgentScheduled string `json:"agentscheduled"`
```
Comment (jhorwit2) [inline pkg/apis/mysql/v1alpha1/types.go]: SheduledAgent looks more correct to me. Also camel case on the json field. 

Or even AgentName

File: pkg/apis/mysql/v1alpha1/types.go [hunk 20]
```diff
@@ -0,0 +1,393 @@
...
+	defaultVersion,
+}
+
+// ClusterSpec defines the attributes a user can specify when creating a cluster
+type ClusterSpec struct {
+	// Version defines the MySQL Docker image version.
+	Version string `json:"version"`
+
+	// Replicas defines the number of running MySQL instances in a cluster
+	Replicas int32 `json:"replicas,omitempty"`
```
Comment (prydie) [inline pkg/apis/mysql/v1alpha1/types.go]: Controversially (perhaps) I'd like to change this to `Members`. `Replicas` is semantically incorrect for the default and recommended deployment (i.e. single-primary mode). The MySQL documentation discusses cluster "members" and I think we should align with that.

//cc @owainlewis @garthy @jhorwit2

File: Gopkg.toml [hunk 21]
```diff
@@ -49,23 +47,23 @@ required = [
 
 [[constraint]]
   name = "k8s.io/api"
-  branch = "release-1.9"
+  version = "kubernetes-1.10.0"
```
Comment (owainlewis) [inline Gopkg.toml]: Dependencies need approvals.

File: docs/user/clusters.md [hunk 22]
```diff
@@ -130,12 +130,12 @@ spec:
...
 ---
 apiVersion: mysql.oracle.com/v1alpha1
-kind: MySQLCluster
+kind: Cluster
 metadata:
   name: example-mysql-cluster-with-volume
 spec:
-  replicas: 1
-  secretRef:
+  mem   : 1
```
Comment (owainlewis) [inline docs/user/clusters.md]: Typo => members

File: pkg/apis/mysql/v1alpha1/helpers.go [hunk 23]
```diff
@@ -0,0 +1,162 @@
...
+	// ClusterNameMaxLen is the maximum supported length of a
+	// Cluster name.
+	// See: https://bugs.mysql.com/bug.php?id=90601
+	ClusterNameMaxLen = 28
+)
+
+// TODO (owain) we need to remove this because it's not reasonable for us to maintain a list
+// of all the potential MySQL versions that can be used and in reality, it shouldn't matter
+// too much. The burden of this is not worth the benfit to a user
+var validVersions = []string{
```
Comment (owainlewis) [inline pkg/apis/mysql/v1alpha1/helpers.go]: Can we now remove this?

**Comment by prydie** [comment]
@gianlucaborello I’d be very interested on your thoughts on these API changes if you have a moment to take a look at the updated type definitions.

**Comment by gianlucaborello** [comment]
Took a quick look (without trying it though) and all seems fine to me, overall minor changes from an API perspective. At the moment, as my RFC PRs hint at, my goal is to make sure the cluster membership management logic is as bullet proof as it can be.

Thanks.

File: Gopkg.toml (hunk 21 above)
Comment (owainlewis) [reply Gopkg.toml]: Done

**Pull request merged**